kitsune/wiki/handlers.py

```diff
@@ -11,30 +12,41 @@ class DocumentListener(UserDeletionListener):
     """Listener for document-related tasks."""
 
     def on_user_deletion(self, user: User) -> None:
-        """Handle the deletion of a user."""
+        """Handle the deletion of a user.
 
-        documents = Document.objects.filter(contributors=user)
+        - Replace the user as a document contributor with content group members
+          if they were the only contributor.
+        - Anonymize the user's revision votes.
+        - Delete the user's non-approved revisions.
+        - Replace the user as creator/reviewer of approved revisions with the SUMO bot.
+        """
+        sumo_bot = Profile.get_sumo_bot()
         content_group = Group.objects.get(name=settings.SUMO_CONTENT_GROUP)
+
+        # Handle approved revisions first to avoid cascade deletion issues
+        Revision.objects.filter(creator=user, is_approved=True).update(creator=sumo_bot)
+
+        # Update reviewer and readied_for_localization_by fields
+        Revision.objects.filter(Q(reviewer=user) | Q(readied_for_localization_by=user)).update(
+            reviewer=sumo_bot, readied_for_localization_by=sumo_bot
+        )
+
+        # Update the list of documents that will need contributor updates
+        documents = Document.objects.filter(contributors=user)
         for document in documents:
             if not document.contributors.exclude(id=user.id).exists():
                 document.contributors.add(*content_group.user_set.all())
             document.contributors.remove(user)
 
-        non_approved_revisions = Revision.objects.filter(creator=user, is_approved=False)
-        # Delete documents with no approved revisions
-        Document.objects.filter(
-            current_revision__isnull=True, revisions__in=non_approved_revisions
-        ).delete()
-        # delete remaining non-approved revisions
-        non_approved_revisions.delete()
-
-        sumo_bot = Profile.get_sumo_bot()
-        Revision.objects.filter(creator=user).update(creator=sumo_bot)
-        Revision.objects.filter(reviewer=user).update(reviewer=sumo_bot)
-        Revision.objects.filter(readied_for_localization_by=user).update(
-            readied_for_localization_by=sumo_bot
-        )
-        # Anonymize any revision votes.
+        # Anonymize the user's votes
         HelpfulVote.objects.filter(creator=user).update(
             creator=None, anonymous_id=AnonymousIdentity().anonymous_id
         )
+
+        # Finally, delete non-approved revisions
+        Document.objects.filter(
+            current_revision__isnull=True,
+            revisions__creator=user,
+            revisions__is_approved=False,
```

Comment:
This `revisions__is_approved=False` is not needed, because it's already covered by the `current_revision__isnull=True` -- or, we could keep it, and get rid of the `current_revision__isnull=True`.